feat(devex): hogli doctor:disk cleanup

webjunkie opened the pull request:
**Summary**

Add `hogli doctor:disk` - interactive disk cleanup tool for PostHog dev environments.

**Potential savings**
- Flox logs: 32GB+ (older than 7 days)
- Git repo: 1-1.5GB (25-40% of .git via gc --aggressive)
- Rust targets: 400MB+ each (feature flag builds scattered repo-wide)
- Python caches: Variable (__pycache__, mypy, pytest, ruff)
- Docker: Variable (unused images, containers, volumes)
- Node artifacts: Variable (build caches, dist folders)
- pnpm store: Variable (unreferenced packages)

**Usage**
```bash
hogli doctor:disk                    # Interactive, all areas
hogli doctor:disk --dry-run          # Preview
hogli doctor:disk --area=git --yes   # Auto git cleanup
hogli doctor:disk --area=flox-logs --area=rust --yes
```

**Available areas**
- `flox-logs` - Flox logs >7 days (via find)
- `git` - Git prune + gc --aggressive  
- `rust` - Cargo targets (repo-wide scan)
- `python` - Python cache directories
- `docker` - Docker system prune
- `dagster` - Dagster storage >7 days
- `node-artifacts` - Build caches
- `pnpm-store` - Prune global store

**Key decisions**
- Use binary units (KiB/MiB/GiB) for accurate size reporting
- Simple find command for Flox logs (not complex Python scanning)
- Replaced node_modules deletion with pnpm store prune (no reinstall needed)
- Skip prompts in dry-run mode automatically
- Git cleanup handles worktrees correctly

webjunkie committed:
feat(hogli): add doctor:disk interactive cleanup command

Add interactive disk space cleanup tool to help developers reclaim disk space
from common PostHog dev bloat.

Cleanup categories:
- Flox logs (.flox/log/) - can grow to 32GB+
- Docker system (unused images, containers, build cache)
- Python caches (__pycache__, .mypy_cache, .pytest_cache)
- Dagster storage (runs/logs older than 7 days)
- Node.js caches (.parcel-cache, .eslintcache, etc.)
- node_modules (requires pnpm install after)

Features:
- Interactive Y/n prompts for each category
- Shows estimated sizes and Docker's native df output
- --dry-run flag to preview without deleting
- --yes flag to auto-confirm all operations
- Graceful error handling for missing tools/directories

Based on findings from PR #39682 where Flox logs reached 32GB.

webjunkie committed:
fix: correct size units and improve Rust target cleanup

- Use binary units (KiB, MiB, GiB) instead of metric (KB, MB, GB) to match 1024-based division
- Scan entire repository for Rust target directories, not just rust/ subdirectory
- Add validation to ensure detected directories are actually Cargo targets
- Update descriptions to reflect actual severity (32GB+ Flox logs, 400MB+ per Rust target)
- Add note about Flox's existing 7-day auto-cleanup in manifest

Fixes issue where feature flag debug builds scattered throughout the repo weren't being detected.

webjunkie committed:
feat(hogli): add --flox flag and 7-day log filter

- Add --flox flag for non-interactive, silent cleanup (used by Flox manifest auto-cleanup)
- Change Flox logs cleanup to only remove files older than 7 days (matching Flox's behavior)
- Update Flox manifest to use `hogli doctor:disk --flox` instead of direct find command
- Silent mode suppresses all output and summary when --flox is used
- Fallback to direct find command if hogli isn't available yet

This makes the cleanup composable: Flox auto-cleanup on activation uses the same
logic as manual cleanup, with consistent 7-day filtering.

webjunkie committed:
feat(hogli): add category-specific flags for doctor:disk

Add individual flags to target specific cleanup categories:
- --flox-logs: Clean Flox logs only
- --docker: Clean Docker system only
- --python: Clean Python caches only
- --dagster: Clean Dagster storage only
- --node-artifacts: Clean Node.js/TS build artifacts only
- --rust: Clean Rust Cargo targets only
- --node-modules: Clean node_modules only

Without flags, all categories run interactively (existing behavior).
With any flag, only specified categories run.

Examples:
  hogli doctor:disk --rust --python    # Clean only Rust and Python
  hogli doctor:disk --flox             # Silent, auto Flox logs cleanup
  hogli doctor:disk                    # Interactive, all categories

This makes the tool composable and allows targeted cleanup operations.

webjunkie committed:
refactor(hogli): simplify doctor:disk to use --area flag only

Remove `--flox` flag in favor of explicit `--area` usage.

Interface:
  --area=<name>  Target specific cleanup area(s) (can be repeated)
  --yes, -y      Auto-confirm all operations
  --dry-run      Preview without deleting

Available areas:
- flox-logs: Flox logs older than 7 days
- docker: Docker system
- python: Python caches
- dagster: Dagster storage
- node-artifacts: Node.js/TS build artifacts
- rust: Rust Cargo targets
- node-modules: pnpm node_modules

Usage:
  hogli doctor:disk                           # Interactive, all areas
  hogli doctor:disk --area=rust --area=python # Specific areas
  hogli doctor:disk --area=flox-logs --yes    # Auto Flox logs (for manifest)
  hogli doctor:disk --area=rust --dry-run     # Preview

Updated Flox manifest to: hogli doctor:disk --area=flox-logs --yes

Simpler interface with explicit, composable flags.

webjunkie committed:
refactor(hogli): move doctor:disk to separate module

Move all doctor:disk implementation from commands.py to doctor.py for better organization.

- common/hogli/doctor.py: All disk cleanup implementation
- common/hogli/commands.py: Import from doctor module to register commands

No functional changes, pure refactor for code cleanliness.

webjunkie committed:
refactor(hogli): simplify Flox logs cleanup with find command

- Restore example command in commands.py for reference
- Replace complex Python log scanning with simple find command
- Flox logs cleanup now runs: find .flox/log -name '*.log' -mtime +7 -delete
- Keep hogli wrapper for future iteration without manifest changes

Simpler is better - one line find command vs dozens of Python lines.

webjunkie committed:
fix(hogli): skip confirmation prompts in dry-run mode

In dry-run mode, auto-proceed through all categories without prompting
for confirmation. Users shouldn't need to answer Y/n for each category
when just previewing what would be cleaned.

webjunkie committed:
refactor(hogli): replace node_modules deletion with pnpm store prune

Replace the nuclear `node_modules` deletion with `pnpm store prune` - a safe,
efficient cleanup that doesn't require reinstalling dependencies.

Why:
- Deleting node_modules is an anti-pattern for cleanup
- Requires full reinstall (slow, wasteful)
- Loses entire dependency graph
- Masks underlying issues

Better approach:
- `pnpm store prune` removes unreferenced packages from global store
- No reinstall needed
- Safe, fast, and actually reclaims space where it accumulates
- Follows package manager best practices

Changed:
- --area=node-modules → --area=pnpm-store
- Removed _collect_node_modules function
- Added _estimate_pnpm_store and _cleanup_pnpm_store

webjunkie committed:
fix(hogli): always prompt for Flox logs cleanup

Set skip_if_empty=False for Flox logs category so it prompts even when
we don't calculate exact sizes (since we use find command). Otherwise
it would skip silently when total_size=0.

webjunkie committed:
feat(hogli): add Git cleanup (prune + gc --aggressive)

Add comprehensive Git cleanup that combines Level 1 + Level 2 operations:
1. git remote prune origin - Remove stale remote branch refs
2. git reflog expire --expire=now --all - Clean old reflogs
3. git gc --prune=now --aggressive - Repack and remove unreachable objects

Expected savings: 25-40% of .git size (1-1.5GB in PostHog's 3.8GB repo)

Features:
- Handles git worktrees by detecting and using main .git directory
- Shows pack count and current size in estimate
- Progress messages during cleanup (gc can take 1-2 minutes)
- Safe, non-destructive operations

Usage: hogli doctor:disk --area=git --yes

webjunkie committed:
refactor(hogli): simplify git cleanup to show native output

Remove custom size calculation - just let git commands show their
native stdout/stderr. Git gc already shows progress and statistics,
no need to parse and reformat it.

Simpler and users see the actual git output.

webjunkie committed:
Merge remote-tracking branch 'origin/master' into codex/analyze-hogli-doctor-disk-command

github-actions[bot] committed:
Update UI snapshots (all browsers)

webjunkie committed:
feat(hogli): use cargo clean for rust cleanup in doctor:disk

Replace manual target directory deletion with proper cargo clean command.

Changes:
- Add _find_cargo_workspaces() to detect all Cargo workspace roots
- Add _cleanup_rust() to run cargo clean in each workspace
- Update _estimate_rust_targets() to check cargo availability
- Set include_in_total=False, skip_if_empty=False for consistency with other external commands
- Skip .flox directories to avoid cleaning dependencies

Cargo clean is more robust than manual deletion as it properly handles all build artifacts and cache files according to Cargo's own cleanup logic.

webjunkie committed:
Merge remote-tracking branch 'origin/master' into codex/analyze-hogli-doctor-disk-command

joshsny commented on common/hogli/doctor.py:
```diff
@@ -0,0 +1,922 @@
+    "pnpm-debug.log",
+    "frontend/.cache",
+    "frontend/tmp",
+    "frontend/types",
```
Why are we cleaning up the types folder?

joshsny reviewed (approved):
Looks great - if it's working locally for you lets ship it and have a few engineers try it out - thanks for doing this!

webjunkie committed:
chore: remove accidentally committed backup file

webjunkie committed:
fix: address code review feedback

- Fix string concatenation style (remove double space)
- Remove unused _iter_named_directories function
- Add explanatory comment to empty except clause
- Add directory existence check in _cleanup_flox_logs
- Remove non-existent frontend/types from cleanup patterns

webjunkie committed:
Merge branch 'master' into codex/analyze-hogli-doctor-disk-command
